metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -225,6 +272,32 @@ private ApiError updateFeature(
         if (featureName.equals(MetadataVersion.FEATURE_NAME)) {
             // Perform additional checks if we're updating metadata.version
             return updateMetadataVersion(newVersion, upgradeType.equals(FeatureUpdate.UpgradeType.UNSAFE_DOWNGRADE), records::add);
+        } else if (featureName.equals(KRaftVersion.FEATURE_NAME)) {
+            if (upgradeType.equals(FeatureUpdate.UpgradeType.UPGRADE)) {
+                try {
+                    kraftVersionAccessor.upgradeKRaftVersion(
+                        currentClaimedEpoch,
+                        KRaftVersion.fromFeatureLevel(newVersion),
```

Comment:
you need to handle the `RuntimeException` that will be thrown if `newVersion` cannot be turned into a `KRaftVersion` object. Otherwise you will cause a controller failover here potentially.

Reply:
Fixed. Changed the code to throw `IllegalArgumentException`. This also matches what the code in `Feature` does.